### PR TITLE
Migrations mutex

### PR DESCRIPTION
**Pull request by vanruch**
# Checklist:

- [x] Code follows the style guidelines
- [x] Documentation updated
- [x] Corner cases evaluated and handled
- [x] Automatic tests added/update
- [x] Automatic tests pass
- [x] Software runs (locally or hosted)
- [x] Sanity check passed
- [x] Self-review passed

### CODE REVIEW

File: src/migrations/Migrator.js [hunk 1]
```diff
@@ -51,22 +52,43 @@ class Migrator {
...
+      }
+      logger.info('Another migration is running. Waiting for it to end');
+      ({migrationStartTime} = (await this.db.collection('migrations').findOne({})));
+      await sleep(this.config.migrationSleepTime);
+    } while (migrationStartTime + this.config.migrationTaskTimeout > getTimestamp());
+  }
+
+  async initMigrations(logger) {
     await this.db.createCollection('migrations');
     const result = await this.db.collection('migrations').findOne({});
```
Comment (spherefoundry) [inline src/migrations/Migrator.js]: As much of this as possible should by covered by the mutex

Comment (vanruch) [reply src/migrations/Migrator.js]: 👍

Comment (vanruch) [reply src/migrations/Migrator.js]: Could not cover it by mutex (creates chicken-egg problem), so just collection creation to be done atomically

File: src/migrations/Migrator.js [hunk 2]
```diff
@@ -51,22 +51,40 @@ class Migrator {
...
   }
 
-  async initMigrations() {
-    await this.db.createCollection('migrations');
-    const result = await this.db.collection('migrations').findOne({});
-    if (!result) {
-      await this.db.collection('migrations').insertOne({version: 0});
-    }
+  async waitForOpenWindowAndStartMigration(logger) {
+    const sleep = async (timeout) => new Promise((resolve) => {
```
[the remark this answers is not in the document]
Comment (vanruch) [reply src/migrations/Migrator.js]: :+1:

File: src/migrations/Migrator.js [hunk 3]
```diff
@@ -19,8 +20,8 @@ class Migrator {
 
   getVersionFromPath(name) {
     const filename = path.basename(name);
-    const splitted = filename.split('_');
-    return parseInt(splitted[0], 10);
+    const split = filename.split('_');
```
[the remark this answers is not in the document]
Comment (vanruch) [reply src/migrations/Migrator.js]: 👍

File: src/migrations/Migrator.js [hunk 4]
```diff
@@ -51,22 +51,56 @@ class Migrator {
...
+      if (modifiedCount > 0) {
+        return;
+      }
+      logger.info('Another migration is running. Waiting for it to end.');
+      await sleep(this.config.migrationSleepTimeInSeconds);
+      ({migrationStartTime} = (await this.db.collection('migrations').findOne({})));
+    } while (this.anotherMigrationIsNotHanged(migrationStartTime));
+  }
+
+  anotherMigrationIsNotHanged(migrationStartTime) {
```
[the remark this answers is not in the document]
Comment (vanruch) [reply src/migrations/Migrator.js]: 👍

File: test/migrations/migrator.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    });
+
+    it('waits for migration task timeout when another migration is running', async () => {
+      await migrator.initMigrations(logger);
+      let callCount = 0;
+      await migrator.waitForOtherMigrationsAndMarkAsStarted(logger, () => {
+        callCount++;
+        clock.tick(migrationSleepTimeInSeconds * 1000);
+      });
+      expect(callCount).to.equal(migrationTaskTimeoutInSeconds);
```
[the remark this answers is not in the document]
Comment (vanruch) [reply test/migrations/migrator.js]: 👍

File: test/migrations/migrator.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+    it('waits until the other migration is finished', async () => {
+      await migrator.initMigrations(logger);
+      let callCount = 0;
+      await migrator.waitForOtherMigrationsAndMarkAsStarted(logger, async () => {
+        callCount++;
+        await migrator.markMigrationAsDone();
+        clock.tick(migrationSleepTimeInSeconds * 1000);
+      });
+      expect(callCount).to.equal(1);
```
[the remark this answers is not in the document]
Comment (vanruch) [reply test/migrations/migrator.js]: :+1:

**Comment by vanruch** [comment]
@spherefoundry CR fixes done

File: test/migrations/migrator.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  before(async () => {
+    ({db, client} = await connectToMongo(config));
+    migrator = new Migrator(db, {
+      migrationSleepTimeInSeconds
+    });
+    logger = new EmptyLogger();
+  });
+
+  beforeEach(() => {
+    clock = sinon.useFakeTimers(now * 1000);
```
Comment (spherefoundry) [inline test/migrations/migrator.js]: This is no longer needed as you can now inject a sleep function that resolves immediately. 

`sinon.useFakeTimers` is evil 😈

File: test/migrations/migrator.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+    it('works correctly if migrationRunning field does not exist in the migration document', async () => {
+      await db.collection('migrations').insertOne({version: 123});
+      await migrator.initMigrations(logger);
+      expect(await db.collection('migrations').findOne({})).to.include({version: 123, migrationRunning: true});
+    });
+
+    it('waits until the other migration is finished', async () => {
+      const expectedIterationsCount = 5;
+      migrator.sleepFunction = async () => {
```
Comment (spherefoundry) [inline test/migrations/migrator.js]: IMO: you should always use the constructor to inject dependencies. You don't know what else has been done to it in the constructor.

File: test/migrations/migrator.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        callCount++;
+        if (callCount === expectedIterationsCount) {
+          await migrator.markMigrationAsDone();
+        }
+        clock.tick(migrationSleepTimeInSeconds * 1000);
```
Comment (spherefoundry) [inline test/migrations/migrator.js]: Also: this is no longer needed. The function can resolve immediately.

File: test/utils/web3_tools.js [hunk 10]
```diff
@@ -56,10 +56,14 @@ describe('Web3 tools', () => {
     const timeout = 3;
     let clock;
 
-    beforeEach(() => {
+    before(() => {
       clock = sinon.useFakeTimers();
     });
 
+    after(() => {
+      clock.restore();
```
Comment (spherefoundry) [inline test/utils/web3_tools.js]: 👍
